text/3430-unix-socket-ancillary-data-v2.md

```diff
@@ -0,0 +1,896 @@
+    fn control_messages(&self) -> &ControlMessages;
+
+    // copy a control message into the ancillary data; error on out-of-capacity.
+    fn add_control_message(
+        &mut self,
+        control_message: impl Into<ControlMessage<'_>>,
+    ) -> Result<(), AncillaryDataNoCapacity>;
+
+    // Add an `SCM_RIGHTS` control message with given borrowed FDs.
+    fn add_file_descriptors(
+        &mut self,
+        borrowed_fds: &[BorrowedFd<'fd>],
+    ) -> Result<(), AncillaryDataNoCapacity>;
+
+    // Transfers ownership of received FDs to the iterator.
+    fn received_fds(&mut self) -> AncillaryDataReceivedFds<'_>;
+
+    // Obtain a mutable buffer usable as the `msg_control` pointer in a call
+    // to `sendmsg()` or `recvmsg()`.
+    fn control_messages_buf(&mut self) -> Option<&mut [u8]>;
+
+    // Update the control messages buffer length according to the result of
+    // calling `sendmsg()` or `recvmsg()`.
+    fn set_control_messages_len(&mut self, len: usize);
+
+    // Scan the control messages buffer for `SCM_RIGHTS` and take ownership of
+    // any file descriptors found within.
+    unsafe fn take_ownership_of_scm_rights(&mut self);
+}
+
+struct AncillaryDataReceivedFds<'a>;
+
+impl<'a> Iterator for AncillaryDataReceivedFds<'a> {
+    type Item = OwnedFd;
+}
+```
+
+#### `fn AncillaryData::received_fds`
+
+By default an `AncillaryData` has no received FDs, and this method will
+ignore FDs added via `add_file_descriptors`. The iterator holds a mutable
+borrow on the `AncillaryData`, and will close any unclaimed FDs when
+it's dropped.
+
+Internally, the iteration works by scanning the control message buffer
+for `SCM_RIGHTS`. As the iteration proceeds the buffer is mutated to set
+the "taken" FDs to `-1` (a sentinal value for Unix file descriptors).
+
+#### `fn AncillaryData::control_messages_buf`
+
+The returned slice, if not `None`, will (1) be non-empty and (2) have the same
+length as the `control_messages_buf` passed to `AncillaryData::new()`. Any
+portion of the buffer not initialized by `add_control_message` or
+`add_file_descriptors` will be zeroed.
+
+When this method is called the `AncillaryData` will have its control
+messages length cleared, so a subsequent call to `control_messages_buf()`
+returns `None`. If the `AncillaryData` contains received FDs, they will
+be closed.
+
+#### `fn AncillaryData::set_control_messages_len`
+
+Does not take ownership of FDs in any `SCM_RIGHTS` control messages that
+might exist within the new buffer length.
+
+**Panics**:
+ * if `len > control_messages_buf.len()`
+ * if `control_messages_buf()` hasn't been called to clear the length.
+
+The second panic condition means that creating an `AncillaryData` and then
+immediately calling `set_control_messages_len` will panic to avoid potentially
+reading uninitialized data.
+
+Also, calling `set_control_messages_len()` twice without an intervening
+`control_messages_buf()` will panic to avoid leaking received FDs.
```

Comment:
> I don't think panicking prevents leaking file descriptors. A panic could be caught with `catch_unwind()`. In most async runtimes this is already done for all spawned tasks, so when a background tasks panics, the process keeps running.

The panic happens before the buffer length is reset. If the panic (assertion) did not happen, then an implementation of `SendMessage` or `RecvMessage` that called `set_control_messages_len()` in the wrong sequence would implicitly leak all owned FDs in the ancillary data, which may be difficult to detect (vs a panic, which is quite easily detectable).


> I also don't really understand why this doesn't take ownership of the file descriptors. I assume it will set the valid range of the control message, and the drop implementation will close them based on this. Won't it?

This function is called by both the `sendmsg()` and `recvmsg()` code paths. The `sendmsg()` code path does not, and must not, take ownership of `SCM_RIGHTS` in its buffer.